contracts/truefi/TrueRatingAgencyV2.sol

```diff
@@ -0,0 +1,485 @@
+
+        loans[id].prediction[choice] = loans[id].prediction[choice].add(stake);
+        loans[id].votes[msg.sender][choice] = loans[id].votes[msg.sender][choice].add(stake);
+
+        require(stakedTrustToken.transferFrom(msg.sender, address(this), stake));
+        emit Voted(id, msg.sender, choice, stake);
+    }
+
+    /**
+     * @dev Vote YES on a loan by staking TRU
+     * @param id Loan ID
+     * @param stake Amount of TRU to stake
+     */
+    function yes(address id, uint256 stake) external override onlyPendingLoans(id) {
+        vote(id, stake, true);
+    }
+
+    /**
+     * @dev Vote NO on a loan by staking TRU
+     * @param id Loan ID
+     * @param stake Amount of TRU to stake
+     */
+    function no(address id, uint256 stake) external override onlyPendingLoans(id) {
+        vote(id, stake, false);
+    }
+
+    // prettier-ignore
+    /**
+     * @dev Withdraw stake on a loan and remove votes.
+     * Unstaking only allowed for loans that are not Running
+     * @param id Loan ID
+     * @param stake Amount of TRU to unstake
+     */
+    function withdraw(address id, uint256 stake) external override {
+        bool choice = loans[id].votes[msg.sender][true] > 0;
+        LoanStatus loanStatus = status(id);
+
+        require(loans[id].votes[msg.sender][choice] >= stake,
+            "TrueRatingAgencyV2: Cannot withdraw more than was staked");
+
+        uint256 amountToTransfer = stake;
+        uint256 burned = 0;
+
+        // if loan still pending, update total votes
+        if (loanStatus == LoanStatus.Pending) {
+            loans[id].prediction[choice] = loans[id].prediction[choice].sub(stake);
+        }
+
+        // if loan status passed pending state claim TRU reward
+        if (loanStatus >= LoanStatus.Running) {
+            claim(id, msg.sender);
+        }
+
+        // update account votes
+        loans[id].votes[msg.sender][choice] = loans[id].votes[msg.sender][choice].sub(stake);
+
+        // transfer tokens to sender and emit event
+        require(stakedTrustToken.transfer(msg.sender, amountToTransfer));
+        emit Withdrawn(id, msg.sender, stake, amountToTransfer, burned);
+    }
+
+    /**
+     * @dev Internal view to convert values to 8 decimals precision
+     * @param input Value to convert to TRU precision
+     * @return output TRU amount
+     */
+    function toTrustToken(uint256 input) internal pure returns (uint256 output) {
+        output = input.div(TOKEN_PRECISION_DIFFERENCE);
+    }
+
+    /**
+     * @dev Update total TRU reward for a Loan
+     * Reward is divided proportionally based on # TRU staked
+     * chi = (TRU remaining in distributor) / (Total TRU allocated for distribution)
+     * interest = (loan APY * term * principal)
+     * R = Total Reward = (interest * chi * rewardFactor)
+     * @param id Loan ID
+     */
+    modifier calculateTotalReward(address id) {
```

Comment:
we might simplify rewards but that'd TBD. I don't see a huge issue with the way it currently is, except that it's hard to understand haha. We might have a fixed distribution per loan instead, since we will eventually phase out this feature.

Reply:
Let's keep it like this in this PR and change it later. Right now it works the same as the old TRA, so the easiest approach is to copy it over. We'll introduce changes in the future.